### PR TITLE
CxFlow Email Enhancements

### PR DESCRIPTION
**Pull request by MithileshPawar**
By submitting a PR to this repository, you agree to the terms within the [Checkmarx Code of Conduct](https://github.com/checkmarx-ts/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/checkmarx-ltd/open-source-template/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

>Changes done to the email functionality in cxflow

### Testing

> Manual Testing done.

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR (if applicable).  *If documentaiton is a Wiki Update, please indicate desired changes within PR MD Comment*
- [] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used

### CODE REVIEW

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010509&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1390
Informational|51

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

File: src/main/java/com/checkmarx/flow/service/ResultsService.java [hunk 1]
```diff
@@ -256,37 +246,32 @@ void logScanDetails(ScanRequest request, Integer projectId, ScanResults results)
...
+
+        if (results != null && !ScanUtils.empty(results.getLink())) {
+            emailCtx.put("issues", results.getXIssues());
+            emailCtx.put("link", results.getLink());
         }
+        emailCtx.put("repo", request.getRepoUrl());
+        emailCtx.put("repo_fullname", namespace.concat("/").concat(repoName));
+        emailService.sendmail(request.getEmail(), message, emailCtx, "template-demo.html");
+        log.info ("Email notification sent.");
     }
```
Comment (AvivCX) [inline src/main/java/com/checkmarx/flow/service/ResultsService.java:275]: email logic should be in email service and not in result service

File: src/main/java/com/checkmarx/flow/service/SastScanner.java [hunk 2]
```diff
@@ -353,13 +353,14 @@ private String getRepoUrl(ScanRequest request, File cxFile) {
...
-                .concat(request.getRepoUrl()));
-        emailCtx.put("heading", "Scan Request Submitted");
-        emailService.sendmail(request.getEmail(), "Checkmarx Scan Submitted for ".concat(request.getNamespace()).concat("/").concat(request.getRepoName()), emailCtx, "message.html");
+        if (flowProperties.getMail() != null && flowProperties.getMail().isNotificationEnabled()) {
+            Map<String, Object> emailCtx = new HashMap<>();
+            emailCtx.put("message", "Checkmarx Scan has been submitted for "
+                    .concat(request.getNamespace()).concat("/").concat(request.getRepoName()).concat(" - ")
+                    .concat(request.getRepoUrl()));
+            emailCtx.put("heading", "Scan Request Submitted");
+            emailService.sendmail(request.getEmail(), "Checkmarx Scan Submitted for ".concat(request.getNamespace()).concat("/").concat(request.getRepoName()), emailCtx, "message.html");
```
Comment (AvivCX) [inline src/main/java/com/checkmarx/flow/service/SastScanner.java:362]: it's better to move the structure of the email to the email service

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

File: src/main/java/com/checkmarx/flow/service/ResultsService.java (hunk 1 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/ResultsService.java]: Done! All email related logic is now in one class.

File: src/main/java/com/checkmarx/flow/service/SastScanner.java (hunk 2 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/SastScanner.java]: Done! All email related logic is now in one class.

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010511&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1390
Informational|51

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

File: src/main/java/com/checkmarx/flow/config/FlowConfig.java [hunk 3]
```diff
@@ -66,6 +68,26 @@ public JavaMailSender getJavaMailSender() {
...
+    public TemplateEngine getTemplateEngine() {
+
+        TemplateEngine templateEngine = new TemplateEngine();
+        templateEngine.setTemplateResolver(fileTemplateResolver());
+
+        return templateEngine;
+
+    }
+
+    public FileTemplateResolver fileTemplateResolver() {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/config/FlowConfig.java:80]: 1) It's better if method name contains an action (e.g. getFileTemplateResolver).
2) Can it be private?

File: src/main/java/com/checkmarx/flow/config/FlowConfig.java [hunk 4]
```diff
@@ -66,6 +68,26 @@ public JavaMailSender getJavaMailSender() {
...
+        FileTemplateResolver fileTemplateResolver = new FileTemplateResolver();
+        fileTemplateResolver.setSuffix(".html");
+        fileTemplateResolver.setTemplateMode(TemplateMode.HTML);
+        fileTemplateResolver.setCharacterEncoding("UTF-8");
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/config/FlowConfig.java:84]: java.nio.charset.StandardCharsets.UTF_8 can be used instead

File: src/main/java/com/checkmarx/flow/config/FlowProperties.java [hunk 5]
```diff
@@ -311,7 +311,9 @@ public void setThresholds(Map<FindingSeverity, Integer> thresholds) {
         private Integer port = 25;
         private String username;
         private String password;
-        private boolean enabled = false;
+        private List<String> cc;
+        private boolean notification = false;
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/config/FlowProperties.java:315]: isNotificationEnabled or notificationEnabled would be a better name.

File: src/main/java/com/checkmarx/flow/service/EmailService.java [hunk 6]
```diff
@@ -13,20 +16,24 @@
...
 import java.util.List;
 import java.util.Map;
 
 @Service
 public class EmailService {
 
+    public static final String MESSAGE_KEY = "message";
+    public static final String HEADING_KEY = "heading";
+    public static final String COMPLETED_PROCESSING = "Successfully completed processing for ";
     private static final Logger log = org.slf4j.LoggerFactory.getLogger(EmailService.class);
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/EmailService.java:29]: @Slf4j annotation of Lombok can be used instead.

File: src/main/java/com/checkmarx/flow/service/EmailService.java [hunk 7]
```diff
@@ -39,34 +46,36 @@ public EmailService(FlowProperties properties, TemplateEngine templateEngine, Ja
...
+
         MimeMessagePreparator messagePreparator = mimeMessage -> {
-            log.info("Sending email");
+            log.info("Sending email notification.");
             MimeMessageHelper messageHelper = new MimeMessageHelper(mimeMessage);
 
-            String contact = properties.getContact();
-            if(!ScanUtils.empty(contact)){
+            String contact = flowProperties.getContact();
+            if (!ScanUtils.empty(contact)) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/EmailService.java:56]: It's better to use org.apache.commons.lang3.StringUtils#isEmpty

File: src/main/java/com/checkmarx/flow/service/EmailService.java [hunk 8]
```diff
@@ -39,34 +46,36 @@ public EmailService(FlowProperties properties, TemplateEngine templateEngine, Ja
...
                 messageHelper.setFrom(contact);
             }
-            messageHelper.setTo(recipients.toArray(new String[0]));
+
+            if (!ScanUtils.empty(recipients)) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/EmailService.java:60]: It's better to use org.apache.commons.collections.CollectionUtils#isNotEmpty

File: src/main/java/com/checkmarx/flow/service/EmailService.java [hunk 9]
```diff
@@ -39,34 +46,36 @@ public EmailService(FlowProperties properties, TemplateEngine templateEngine, Ja
...
+                messageHelper.setTo(recipients.toArray(new String[0]));
+                messageHelper.setCc((flowProperties.getMail().getCc()).toArray(new String[0]));
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/EmailService.java:62]: Please extract (flowProperties.getMail().getCc()).toArray(new String[0]) into a variable (ctrl+alt+v).

File: src/main/java/com/checkmarx/flow/service/EmailService.java [hunk 10]
```diff
@@ -39,34 +46,36 @@ public EmailService(FlowProperties properties, TemplateEngine templateEngine, Ja
...
+
             messageHelper.setSubject(subject);
             String content = generateContent(ctx, template);
             messageHelper.setText(content, true);
         };
         try {
-            FlowProperties.Mail mail = properties.getMail();
-            if(mail != null && mail.isEnabled()) {
+            FlowProperties.Mail mail = flowProperties.getMail();
+            if (mail != null) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/EmailService.java:73]: Why was isEnabled removed from the condition?

File: src/main/java/com/checkmarx/flow/service/EmailService.java [hunk 11]
```diff
@@ -81,5 +90,65 @@ private String generateContent(Map<String, Object> ctx, String template) {
         context.setVariables(ctx);
         return templateEngine.process(template, context);
     }
+
+    public void sendScanSubmittedEmail(ScanRequest request) {
+        if (isEmailNotificationAllowed()) {
+            String scanSubmittedSubject = "Checkmarx Scan Submitted for ".concat(request.getNamespace()).concat("/").concat(request.getRepoName());
+            String scanSubmittedMessage = "Checkmarx Scan has been submitted for ".concat(request.getNamespace()).concat("/").concat(request.getRepoName())
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/EmailService.java:97]: It's more readable with String.format().

File: src/main/java/com/checkmarx/flow/service/SastScanner.java [hunk 12]
```diff
@@ -378,7 +371,7 @@ private Integer handleUnKnownProjectId(Integer projectId, String ownerId, String
 
     private void checkScanSubmitEmailDelivery(ScanRequest scanRequest) {
         if (!ScanUtils.anyEmpty(scanRequest.getNamespace(), scanRequest.getRepoName(), scanRequest.getRepoUrl())) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/SastScanner.java:373]: org.apache.commons.lang3.StringUtils#isNoneEmpty can be used instead.

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

File: src/main/java/com/checkmarx/flow/config/FlowConfig.java (hunk 3 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/config/FlowConfig.java]: Done!

File: src/main/java/com/checkmarx/flow/config/FlowProperties.java (hunk 5 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/config/FlowProperties.java]: Renamed to notificationEnabled.

File: src/main/java/com/checkmarx/flow/service/EmailService.java (hunk 7 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/EmailService.java]: Done.

File: src/main/java/com/checkmarx/flow/service/EmailService.java (hunk 9 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/EmailService.java]: Done.

File: src/main/java/com/checkmarx/flow/service/EmailService.java (hunk 11 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/EmailService.java]: Changed to use String.format()

File: src/main/java/com/checkmarx/flow/service/SastScanner.java (hunk 12 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/SastScanner.java]: Done.

File: src/main/java/com/checkmarx/flow/service/EmailService.java (hunk 10 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/EmailService.java:77]: isEnabled was removed to keep the functionality separate for usual PR notification mails and Email as a bug tracker. 
Appropriate conditions are used at the point where the mails are triggered. So this condition was being redundant.

File: src/main/java/com/checkmarx/flow/service/EmailService.java (hunk 6 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/EmailService.java]: Done

File: src/main/java/com/checkmarx/flow/config/FlowConfig.java (hunk 4 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/config/FlowConfig.java]: Done

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010538&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1382
Informational|51

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010539&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1382
Informational|51

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010541&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1382
Informational|51

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

File: src/main/java/com/checkmarx/flow/service/EmailService.java (hunk 8 above)
Comment (MithileshPawar) [reply src/main/java/com/checkmarx/flow/service/EmailService.java]: Done

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010543&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|112
Low|1382
Informational|51

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by MithileshPawar** [comment]
Hi @AvivCX , I think one of the component test (GetResultsAnalyticsTestSteps) is failing because the emailService (from ResultService) is null. Can you please ask someone from your team to take a look and confirm?

**Comment by AvivCX** [comment]
@NimrodGolan can you please help Mithilesh to fix this component test?

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010564&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010565&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010566&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
### Checkmarx scan completed
[Full Scan Details](http://EC2AMAZ-JP2I9QF/CxWebClient/ViewerMain.aspx?scanid=1010567&projectid=175)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|0
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

#### Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

**Comment by MithileshPawar** [comment]
@NimrodGolan , @AvivCX thanks for your support, the tests are all green now.

**Pull request merged**